File: articles.jdbc/src/test/java/pet/db/jdbc/service/ReviewServiceTest.java
package pet.db.jdbc.service;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import pet.db.jdbc.entity.Article;
import pet.db.jdbc.entity.Review;
import pet.db.jdbc.entity.User;
import pet.db.jdbc.tool.db.DbCleaner;
import pet.db.jdbc.tool.generator.TestDataGenerator;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@SpringBootTest
public class ReviewServiceTest {

    @Autowired
    private DbCleaner dbCleaner;

    @Autowired
    private ReviewService reviewService;

    @Autowired
    private TestDataGenerator<User> userTestDataGenerator;

    @Autowired
    private TestDataGenerator<Article> articleTestDataGenerator;

    @Autowired
    private TestDataGenerator<Review> reviewTestDataGenerator;


    @AfterEach
    void cleanDb() {
        dbCleaner.cleanAll();
    }

    @Test
    public void createReview() {
        Review reviewForSave = reviewTestDataGenerator.generateUnsavedData();

        Review savedReview = reviewService.create(reviewForSave);

        Optional<Review> reviewOptionalForCheck = reviewService.findById(savedReview.getId());
        assertTrue(reviewOptionalForCheck.isPresent());
        assertEquals(savedReview, reviewOptionalForCheck.get());
    }

    @Test
    public void createReviewWithInvalidData() {
        Review reviewForSave = reviewTestDataGenerator.generateUnsavedData();
        reviewForSave.setContent(null);

        assertThatThrownBy(() -> reviewService.create(reviewForSave))
                .isInstanceOf(RuntimeException.class);;
    }

    @Test
    public void createReviewWithNonExistentArticleId() {
        Review reviewForSave = reviewTestDataGenerator.generateUnsavedData();
        Article unsavedArticle = articleTestDataGenerator.generateUnsavedData();
        reviewForSave.setArticleId(unsavedArticle.getId());

        assertThatThrownBy(() -> reviewService.create(reviewForSave))
                .isInstanceOf(RuntimeException.class);;
    }

    @Test
    public void createReviewWithNonExistentAuthorId() {
        User unsavedAuthor = userTestDataGenerator.generateUnsavedData();
        Review reviewForSave = reviewTestDataGenerator.generateUnsavedData();
        reviewForSave.setAuthorId(unsavedAuthor.getId());

        assertThatThrownBy(() -> reviewService.create(reviewForSave))
                .isInstanceOf(RuntimeException.class);;
    }

    @Test
    public void deleteReviewById() {
        Review savedReview = reviewTestDataGenerator.generateSavedData();

        reviewService.deleteById(savedReview.getId());

        Optional<Review> reviewOptionalForCheck = reviewService.findById(savedReview.getId());
        assertTrue(reviewOptionalForCheck.isEmpty());
    }

    @Test
    public void deleteReviewByNonExistentId() {
        Review unsavedReview = reviewTestDataGenerator.generateUnsavedData();

        reviewService.deleteById(unsavedReview.getId());

        Optional<Review> reviewOptionalForCheck = reviewService.findById(unsavedReview.getId());
        assertTrue(reviewOptionalForCheck.isEmpty());
    }

    @Test
    public void findReviewById() {
        Review savedReview = reviewTestDataGenerator.generateSavedData();

        Optional<Review> reviewOptionalForCheck = reviewService.findById(savedReview.getId());

        assertTrue(reviewOptionalForCheck.isPresent());
        assertEquals(savedReview, reviewOptionalForCheck.get());
    }

    @Test
    public void findReviewByNonExistentId() {
        Review unsavedReview = reviewTestDataGenerator.generateUnsavedData();

        Optional<Review> reviewOptionalForCheck = reviewService.findById(unsavedReview.getId());

        assertTrue(reviewOptionalForCheck.isEmpty());
    }

    @Test
    public void findReviewsByAuthorId() {
        List<Review> reviews = reviewTestDataGenerator.generateSavedData(30);
        Integer authorId = reviews.getFirst().getAuthorId();
        List<Review> allReviewsByAuthor = reviews.stream()
                .filter(x -> Objects.equals(x.getAuthorId(), authorId))
                .toList();

        List<Review> reviewsForCheck = reviewService.findByAuthorId(authorId);


        assertEquals(allReviewsByAuthor.size(), reviewsForCheck.size());
        assertTrue(allReviewsByAuthor.containsAll(reviewsForCheck));
        assertTrue(reviewsForCheck.containsAll(allReviewsByAuthor));
    }

    @Test
    public void findReviewsByNonExistentAuthorId() {
        User unsavedAuthor = userTestDataGenerator.generateUnsavedData();

        assertThatThrownBy(() -> reviewService.findByAuthorId(unsavedAuthor.getId()))
                .isInstanceOf(NoSuchElementException.class);;
    }

    @Test
    public void findReviewsByArticleId() {
        List<Review> reviews = reviewTestDataGenerator.generateSavedData(30);
        Integer articleId = reviews.getFirst().getArticleId();
        List<Review> allReviewsForArticle = reviews.stream()
                .filter(x -> Objects.equals(x.getArticleId(), articleId))
                .toList();

        List<Review> reviewsForCheck = reviewService.findByArticleId(articleId);

        assertEquals(allReviewsForArticle.size(), reviewsForCheck.size());
        assertTrue(allReviewsForArticle.containsAll(reviewsForCheck));
        assertTrue(reviewsForCheck.containsAll(allReviewsForArticle));
    }

    @Test
    public void findReviewsByNonExistentArticleId() {
        Article unsavedArticle = articleTestDataGenerator.generateUnsavedData();

        assertThatThrownBy(() -> reviewService.findByArticleId(unsavedArticle.getId()))
                .isInstanceOf(NoSuchElementException.class);;
    }

}

File: src/test/kotlin/pet/articles/service/ReviewServiceTest.kt
package pet.articles.service

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.koin.core.qualifier.named
import org.koin.test.KoinTest
import org.koin.test.inject

import pet.articles.model.dto.Article
import pet.articles.model.dto.Review
import pet.articles.model.dto.User
import pet.articles.service.article.ArticleService
import pet.articles.service.review.ReviewService
import pet.articles.service.user.UserService
import pet.articles.tool.generator.TestDataGenerator
import pet.articles.tool.testing.extension.DBCleanupExtension
import pet.articles.tool.testing.extension.KoinConfigureTestExtension

@ExtendWith(DBCleanupExtension::class)
@ExtendWith(KoinConfigureTestExtension::class)
class ReviewServiceTest : KoinTest {

    companion object {
        const val NUM_OF_TEST_REVIEWS = 10
    }

    private val reviewService: ReviewService by inject()

    private val articleService: ArticleService by inject()

    private val userService: UserService by inject()

    private val userGenerator: TestDataGenerator<User> by inject(
        named("UserTestDataGenerator")
    )

    private val articleGenerator: TestDataGenerator<Article> by inject(
        named("ArticleTestDataGenerator")
    )

    private val reviewGenerator: TestDataGenerator<Review> by inject(
        named("ReviewTestDataGenerator")
    )

    @Test
    fun createReview() {
        val reviewForSave: Review = reviewGenerator.generateUnsavedData()

        val savedReview: Review = reviewService.create(reviewForSave)

        val reviewForCheck: Review? = reviewService.findById(savedReview.id!!)
        assertNotNull(reviewForCheck)
        assertEquals(savedReview, reviewForCheck)
    }

    @Test
    fun createReviewWithNonExistentArticleId() {
        val unsavedArticle: Article = articleGenerator.generateUnsavedData()
        val reviewForSave: Review = reviewGenerator.generateUnsavedData().copy(
            articleId = unsavedArticle.id!!
        )

        assertThrows(RuntimeException::class.java) {
            reviewService.create(reviewForSave)
        }
    }

    @Test
    fun createReviewWithNonExistentAuthorId() {
        val unsavedAuthor: User = userGenerator.generateUnsavedData()
        val reviewForSave: Review = reviewGenerator.generateUnsavedData().copy(
            authorId = unsavedAuthor.id!!
        )

        assertThrows(RuntimeException::class.java) {
            reviewService.create(reviewForSave)
        }
    }

    @Test
    fun deleteReviewById() {
        val savedReview: Review = reviewGenerator.generateSavedData()

        reviewService.deleteById(savedReview.id!!)

        val reviewForCheck: Review? = reviewService.findById(savedReview.id!!)
        assertNull(reviewForCheck)
    }

    @Test
    fun deleteReviewByNonExistentId() {
        val unsavedReview: Review = reviewGenerator.generateUnsavedData()

        reviewService.deleteById(unsavedReview.id!!)

        val reviewForCheck: Review? = reviewService.findById(unsavedReview.id!!)
        assertNull(reviewForCheck)
    }

    @Test
    fun deleteReviewWhenDeletingArticle() {
        val savedReview: Review = reviewGenerator.generateSavedData()
        val article: Article = articleService.findById(savedReview.articleId)!!

        articleService.deleteById(article.id!!)

        val reviewForCheck: Review? = reviewService.findById(savedReview.id!!)
        assertNull(reviewForCheck)
    }

    @Test
    fun deleteReviewWhenDeletingAuthor() {
        val savedReview: Review = reviewGenerator.generateSavedData()
        val author: User = userService.findById(savedReview.authorId!!)!!

        userService.deleteById(author.id!!)

        val reviewForCheck: Review? = reviewService.findById(savedReview.id!!)
        assertNotNull(reviewForCheck)
        assertNull(reviewForCheck!!.authorId)
    }

    @Test
    fun findReviewById() {
        val savedReview: Review = reviewGenerator.generateSavedData()

        val reviewForCheck: Review? = reviewService.findById(savedReview.id!!)

        assertNotNull(reviewForCheck)
        assertEquals(savedReview, reviewForCheck)
    }

    @Test
    fun findReviewByNonExistentId() {
        val unsavedReview: Review = reviewGenerator.generateUnsavedData()

        val reviewForCheck: Review? = reviewService.findById(unsavedReview.id!!)

        assertNull(reviewForCheck)
    }

    @Test
    fun findReviewsByAuthorId() {
        val reviews: List<Review> = reviewGenerator.generateSavedData(NUM_OF_TEST_REVIEWS)
        val authorId: Int = reviews.random().authorId!!
        val allReviewsByAuthor: List<Review> = reviews.filter { it.authorId == authorId }

        val reviewsForCheck: List<Review> = reviewService.findByAuthorId(authorId)

        assertEquals(allReviewsByAuthor.size, reviewsForCheck.size)
        assertTrue(allReviewsByAuthor.containsAll(reviewsForCheck))
        assertTrue(reviewsForCheck.containsAll(allReviewsByAuthor))
    }

    @Test
    fun findReviewsByNonExistentAuthorId() {
        val unsavedAuthor: User = userGenerator.generateUnsavedData()

        val reviews: List<Review> = reviewService.findByAuthorId(unsavedAuthor.id!!)

        assertTrue(reviews.isEmpty())
    }

    @Test
    fun findReviewsByArticleId() {
        val reviews: List<Review> = reviewGenerator.generateSavedData(NUM_OF_TEST_REVIEWS)
        val articleId: Int = reviews.random().articleId
        val allReviewsForArticle: List<Review> = reviews.filter { it.articleId == articleId }

        val reviewsForCheck: List<Review> = reviewService.findByArticleId(articleId)

        assertEquals(allReviewsForArticle.size, reviewsForCheck.size)
        assertTrue(allReviewsForArticle.containsAll(reviewsForCheck))
        assertTrue(reviewsForCheck.containsAll(allReviewsForArticle))
    }

    @Test
    fun findReviewsByNonExistentArticleId() {
        val unsavedArticle: Article = articleGenerator.generateUnsavedData()

        val reviews: List<Review> = reviewService.findByArticleId(unsavedArticle.id!!)

        assertTrue(reviews.isEmpty())
    }
}